File: tests/fixtures/vulnerable-go/sql-injection.go
package main

import (
	"database/sql"
	"fmt"
	"net/http"
)

// VULNERABLE: SQL injection via fmt.Sprintf
func getUserByID(db *sql.DB, userID string) error {
	query := fmt.Sprintf("SELECT * FROM users WHERE id = '%s'", userID)
	rows, err := db.Query(query)
	if err != nil {
		return err
	}
	defer rows.Close()
	return nil
}

// VULNERABLE: SQL injection via string concatenation
func searchUsers(db *sql.DB, searchTerm string) error {
	query := "SELECT * FROM users WHERE name LIKE '%" + searchTerm + "%'"
	_, err := db.Exec(query)
	return err
}

// VULNERABLE: SQL injection in prepared statement
func deleteUser(db *sql.DB, userID string) error {
	stmt, err := db.Prepare(fmt.Sprintf("DELETE FROM users WHERE id = %s", userID))
	if err != nil {
		return err
	}
	defer stmt.Close()
	_, err = stmt.Exec()
	return err
}

func handler(w http.ResponseWriter, r *http.Request, db *sql.DB) {
	userID := r.URL.Query().Get("id")
	getUserByID(db, userID)
}

File: tests/fixtures/vulnerable-rust/sql_injection.rs
use sqlx::postgres::PgConnection;

// Vulnerable: SQL injection with format!
async fn get_user_by_id_vulnerable(conn: &mut PgConnection, user_id: &str) -> Result<(), sqlx::Error> {
    let query = format!("SELECT * FROM users WHERE id = '{}'", user_id);
    sqlx::query(&query).execute(conn).await?;
    Ok(())
}

// Vulnerable: SQL injection with string concatenation
async fn delete_user_vulnerable(conn: &mut PgConnection, username: &str) -> Result<(), sqlx::Error> {
    let query = "DELETE FROM users WHERE username = '".to_string() + username + "'";
    sqlx::query(&query).execute(conn).await?;
    Ok(())
}

// Vulnerable: format! in execute
async fn update_email_vulnerable(conn: &mut PgConnection, user_id: i32, email: &str) -> Result<(), sqlx::Error> {
    conn.execute(format!("UPDATE users SET email = '{}' WHERE id = {}", email, user_id)).await?;
    Ok(())
}

// Vulnerable: prepare with format!
async fn prepare_query_vulnerable(conn: &mut PgConnection, table: &str) -> Result<(), sqlx::Error> {
    let stmt = conn.prepare(format!("SELECT * FROM {}", table)).await?;
    Ok(())
}
